### PR TITLE
[#1908] Upgraded webpage plugin

### PR DESCRIPTION
**Pull request by cadecairos**
test on basic template

### CODE REVIEW

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @cadecairos received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/e195d811b9efa6c/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/e195d811b9efa6c/output.txt

Total script time: 0.96 mins

Server is running on http://pr746.butyr.org/templates/basic/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @cadecairos received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/3dd722d7a3935ab/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/3dd722d7a3935ab/output.txt

Total script time: 1.04 mins

Server is running on http://pr746.butyr.org/templates/basic/

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @cadecairos received. Current queue size: 1

Live output at: http://botio.butyr.org:8000/27bee4a07f9f60c/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @cadecairos received. Current queue size: 2

Live output at: http://botio.butyr.org:8000/c81d6145a26921d/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/27bee4a07f9f60c/output.txt

Total script time: 1.00 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/c81d6145a26921d/output.txt

Total script time: 0.73 mins

Server is running on http://pr746.butyr.org/templates/basic/
Run the Test Suite: http://pr746.butyr.org/test/

File: templates/assets/editors/editorhelper.js [hunk 1]
```diff
@@ -22,11 +22,16 @@
           }
         };
 
-    global.EditorHelper.draggable = function( trackEvent, dragContainer, mediaContainer ) {
+    global.EditorHelper.draggable = function( trackEvent, dragContainer, mediaContainer, handleElem, startFn, endFn ) {
```
Comment (k88hudson) [inline templates/assets/editors/editorhelper.js]: This is a lot of parameters, let's move it into an options object.

File: templates/assets/editors/editorhelper.js [hunk 2]
```diff
@@ -35,12 +40,16 @@
       });
     };
 
-    global.EditorHelper.resizable = function( trackEvent, resizeContainer, mediaContainer, handlePositions ) {
+    global.EditorHelper.resizable = function( trackEvent, resizeContainer, mediaContainer, handlePositions, startFn, endFn ) {
```
Comment (k88hudson) [inline templates/assets/editors/editorhelper.js]: Same thing here - also, a note about what each option does would be nice

File: templates/assets/editors/webpage/popcorn.webpage.editor.css [hunk 3]
```diff
@@ -0,0 +1,29 @@
+iframe.disable-pointer-events {
+  pointer-events: none;
+}
+
+.webpage-drag-handle {
```
Comment (k88hudson) [inline templates/assets/editors/webpage/popcorn.webpage.editor.css]: We're going to have to abstract this into something google maps uses too

File: templates/assets/editors/webpage/popcorn.webpage.editor.css [hunk 4]
```diff
@@ -0,0 +1,29 @@
+iframe.disable-pointer-events {
+  pointer-events: none;
```
Comment (k88hudson) [inline templates/assets/editors/webpage/popcorn.webpage.editor.css]: Needs an ignore line.

File: templates/assets/editors/webpage/popcorn.webpage.editor.css [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  top: 15px;
+  left: 15px;
+}
+
+.webpage-drag-handle:hover {
+  opacity: 1;
+}
+
+iframe.webpage-plugin-iframe {
+  border: none;
```
Comment (k88hudson) [inline templates/assets/editors/webpage/popcorn.webpage.editor.css]: Why border: none? Are iframes not set to border:none in the reset?

File: templates/assets/editors/webpage/webpage.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+/*global EditorHelper*/
+
+EditorHelper.addPlugin( "webpage", function( trackEvent ) {
+  var _popcornOptions = trackEvent.popcornTrackEvent,
+      _wrapper = _popcornOptions._wrapper,
+      _iframe = _popcornOptions._iframe,
+      _media = document.getElementById( trackEvent.track._media.target ),
+      handle = document.createElement( "div" );
```
Comment (k88hudson) [inline templates/assets/editors/webpage/webpage.js]: `_handle`

File: templates/assets/editors/webpage/webpage.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+  function start() {
+    _iframe.classList.add( "disable-pointer-events" );
+  }
+
+  function end() {
+    _iframe.classList.remove( "disable-pointer-events" );
+  }
+
+  if ( window.jQuery ) {
```
Comment (k88hudson) [inline templates/assets/editors/webpage/webpage.js]: You don't need to check this

File: templates/assets/plugins/webpage/popcorn.webpage.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+  Popcorn.plugin( "webpage" , {
+    _setup: function( options ) {
+
+      var target = Popcorn.dom.find( options.target || manifest.options.target[ "default" ] ),
+          manifest = options._natives.manifest,
+          wrapper,
+          iframe;
+
+      options._wrapper = wrapper  = document.createElement( "div" );
```
Comment (k88hudson) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: there's extra space after `wrapper`

File: templates/assets/plugins/webpage/popcorn.webpage.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+      target && target.appendChild( wrapper );
+
+      // make src an iframe acceptable string
+      options.src = options.src.replace( /^(https?:)?(\/\/)?/, "//" );
+
+      // make an iframe
+      options._iframe = iframe = document.createElement( "iframe" );
+
+      // not 100 because the scrollbars interfere with draggable and resizeable
```
Comment (k88hudson) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: resizable

File: templates/assets/plugins/webpage/popcorn.webpage.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      iframe.setAttribute( "width", "100%" );
+      iframe.setAttribute( "height", "100%" );
+      iframe.classList.add( "webpage-plugin-iframe" );
+      iframe.id = options.id || Popcorn.guid();
+      iframe.src = options.src;;
```
Comment (k88hudson) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: extra `;`

File: templates/assets/plugins/webpage/popcorn.webpage.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+      wrapper.appendChild( iframe );
+
+      options.toString = function() {
+        return options.src || manifest.options.src[ "default" ];
+      };
+
+    },
+
+    start: function( event, options ){
+      options._wrapper.style.display = "block";
```
Comment (k88hudson) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: Needs transitions

File: templates/basic/config.json [hunk 12]
```diff
@@ -32,6 +32,10 @@
       {
         "type": "repeat",
         "path": "{{baseDir}}templates/assets/plugins/repeat/popcorn.repeat.js"
+      },
+      {
+        "type": "webpage",
+        "path": "{{baseDir}}templates/assets/plugins/webpage/popcorn.webpage.js"
       }
```
Comment (k88hudson) [inline templates/basic/config.json]: Needs icon

File: templates/assets/editors/webpage/popcorn.webpage.editor.css (hunk 5 above)
Comment (cadecairos) [inline templates/assets/editors/webpage/popcorn.webpage.editor.css]: They aren't,  AFAICT

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @cadecairos received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/67a1c3ca50a0c06/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/67a1c3ca50a0c06/output.txt

Total script time: 0.70 mins

Server is running on http://pr746.butyr.org/templates/basic/
Run the Test Suite: http://pr746.butyr.org/test/

File: templates/assets/editors/popup/popup.js [hunk 13]
```diff
@@ -8,7 +8,9 @@ EditorHelper.addPlugin( "popup", function( trackEvent ) {
   media = document.getElementById( trackEvent.track._media.target );
 
   if ( window.jQuery ) {
-    EditorHelper.draggable( trackEvent, _container, media );
-    EditorHelper.resizable( trackEvent, _container, media, "e" );
+    EditorHelper.draggable( trackEvent, _container, media, {} );
```
Comment (k88hudson) [inline templates/assets/editors/popup/popup.js]: you don't need to pass `{}` here, you do `options = options || {}` inside your functions

File: templates/assets/editors/titles/titles.js [hunk 14]
```diff
@@ -9,7 +9,7 @@ EditorHelper.addPlugin( "titles", function( trackEvent ) {
 
   if ( window.jQuery ) {
     if ( trackEvent.popcornOptions.position === "custom" ) {
-      EditorHelper.draggable( trackEvent, _container, media );
+      EditorHelper.draggable( trackEvent, _container, media, {} );
```
Comment (k88hudson) [inline templates/assets/editors/titles/titles.js]: same thing here

File: templates/assets/plugins/webpage/popcorn.webpage.js [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+      wrapper.style.top = normalize( options.top || manifest.options.top[ "default" ], 0, 90 ) + "%";
+      wrapper.style.left = normalize( options.left || manifest.options.left[ "default" ], 0, 90 ) + "%";
+      wrapper.style.height = normalize( options.height || manifest.options.height[ "default" ], 10, 100 ) + "%";
+      wrapper.style.width = normalize( options.width || manifest.options.width[ "default" ], 10, 100 ) + "%";
+      wrapper.style.position = "absolute";
+      wrapper.style.display = "none";
+
+      wrapper.appendChild( iframe );
+      wrapper.classList.add( "off" );
+
```
Comment (k88hudson) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: add `wrapper.classList.add( "popcorn-fade" )`

File: templates/assets/plugins/webpage/popcorn.webpage.js (hunk 11 above)
Comment (k88hudson) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: Remove this line

File: templates/assets/plugins/webpage/popcorn.webpage.js [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    },
+
+    start: function( event, options ){
+      options._wrapper.style.display = "block";
+      options._wrapper.classList.add( "on" );
+      options._wrapper.classList.remove( "off" );
+    },
+
+    end: function( event, options ){
+      options._wrapper.style.display = "none";
```
Comment (k88hudson) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: Remove this line

Comment (cadecairos) [inline templates/assets/plugins/webpage/popcorn.webpage.js]: what about the display = "block" in start?

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/e6cbd9cdab2ef3a/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/e6cbd9cdab2ef3a/output.txt

Total script time: 1.26 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/005a0b2963d019b/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/005a0b2963d019b/output.txt

Total script time: 0.68 mins

Server is running on http://pr746.butyr.org/templates/basic/
Run the Test Suite: http://pr746.butyr.org/test/

File: templates/assets/editors/webpage/webpage.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+  function startFn() {
+    _iframe.classList.add( "editor-disable-pointer-events" );
+  }
+
+  function endFn() {
+    _iframe.classList.remove( "editor-disable-pointer-events" );
+  }
+
+  window.EditorHelper.resizable( trackEvent, _wrapper, _media, {
+    handles: null,
```
Comment (mjschranz) [inline templates/assets/editors/webpage/webpage.js]: Same, handlePositions

You can make it handle if you want like the draggable, but make sure the naming is consistent in all places it's being used.

File: templates/assets/editors/popup/popup.js [hunk 18]
```diff
@@ -9,6 +9,8 @@ EditorHelper.addPlugin( "popup", function( trackEvent ) {
 
   if ( window.jQuery ) {
     EditorHelper.draggable( trackEvent, _container, media );
-    EditorHelper.resizable( trackEvent, _container, media, "e" );
+    EditorHelper.resizable( trackEvent, _container, media, {
+      handles: "e"
```
Comment (mjschranz) [inline templates/assets/editors/popup/popup.js]: again, needs to be handlePositions

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @mjschranz received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/28b427546ee49d6/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/28b427546ee49d6/output.txt

Total script time: 0.63 mins

Server is running on http://pr746.butyr.org/templates/basic/
Run the Test Suite: http://pr746.butyr.org/test/

**Comment by jbuck** [comment]
Staged on master: https://github.com/mozilla/butter/commit/057bd83eee1c1b94c85fe198763b7f5250d98664

**Pull request closed without merging**